### ISSUE TITLE
CLEARTEXT communication to xxx not permitted by network security policy in Android

### ISSUE THREAD

**Issue by akhilsanker** [klarna/react-native-klarna-inapp-sdk#119]
**Describe the bug**
No API calls starting with HTTP (http://203.89.289/api/get)are working in Android, but API calls starting with HTTPS (https://domain.com/api/get) are working. No issue in iOS.

**To Reproduce**
Steps to reproduce the behaviour:
API calls starting with HTTP are not working at all after installing this package. Even though I got this error from rn-fetch-blob, none of the APIs is working.

**Expected behaviour**
All the API calls starting with HTTP needs to work in Android.

**Device and version:**
 - Android Emulator, One Plus
 - Android 10
 - 30

**Comment by NMGuner** [comment]
Hi @akhilsanker, what URLs are you having troubles with exactly, and are they related to the Klarna Payments? 

As an SDK we absolutely shouldn't enforce any cleartext network setting as HTTP calls are not used by us and starting with Android 9 clear text is disabled by default. But if your application needs these settings for other reasons, can you confirm that you have the `network_security_config.xml` at app level which allows it. If the setting exists and the SDK is somehow overriding that, then we can look into it. If you need further information, please ask us or refer to the [official documentation](https://developer.android.com/training/articles/security-config#CleartextTrafficPermitted).

**Comment by akhilsanker** [comment]
Hi @NMGuner,

Thanks for the response.

Actually, our development serves API URLs are not SSL secured but our production serves API URLs are SSL secured.
After installing the plugin, none of the development server APIs (ex: login, registration, etc..) is working but all the production server APIs are working.

We have already added `android:usesCleartextTraffic="true" ` in the Android manifest and don't have a `network_security_config.xml` file in the Android folder. We use `axios` for network calls.

Is there anything that I am missing to consider while installing the plugin?

Thanks

**Comment by NMGuner** [comment]
That makes total sense and explains why this is happening. The manifest setting is a deprecated solution for configuring this as far as I remember and due to our SDK having a `network_security_config.xml`(which will be applied to the app even if it doesn't have the file itself) that setting in the manifest will be ignored. We can review our network config but I would also suggest you to add a config file to the app instead of the manifest property, which would be a more secure and faster solution for the time being. Other than this I don't think there is anything specific about `axios` or installing the plugin.

It would be great if you could give this a try and let us know if it fixes the issue or not.

**Comment by akhilsanker** [comment]
Hi @NMGuner,

Thanks for the response. I will try the solution and let you know the result.

Thanks

**Comment by akhilsanker** [comment]
Hi @NMGuner,

I have tried adding `network_security_config.xml` in Android folder. But after that the metro bundler is not working and so unable to run the project.

<img width="393" alt="Screenshot 2021-12-14 at 11 16 38 AM" src="https://user-images.githubusercontent.com/32122016/145940511-815a039c-f869-4b6b-bbfb-517ef28ae34d.png">

Metro bundler is still running but there is issue in establishing connection with emulator.

My `network_security_config.xml` is as follows:
```
<?xml version="1.0" encoding="utf-8"?>
<network-security-config>
    <domain-config cleartextTrafficPermitted="true">
        <domain includeSubdomains="true">http://xxx/</domain>
        <domain includeSubdomains="true">http://localhost:8081/</domain>
    </domain-config>
</network-security-config>
```

and in Android manifest added `android:networkSecurityConfig="@xml/network_security_config"` and removed `android:usesCleartextTraffic="true"`

And installing SSL in development serve will be another solution for me will be a workaround I guess, isn't?

Thanks

**Comment by NMGuner** [comment]
Hi @akhilsanker,

That's due to metro server running locally on your machine and requiring a clear text communication, which I believe would only affect your debug builds. There is an issue about this [here](https://github.com/facebook/react-native/issues/22375) (also a [PR](https://github.com/facebook/react-native/pull/23105)) with few presented solutions around it using network security config.

**Comment by akhilsanker** [comment]
@NMGuner,

But the issue with metro bundler is while adding `network_security_config.xml`, as you suggested

> That makes total sense and explains why this is happening. The manifest setting is a deprecated solution for configuring this as far as I remember and due to our SDK having a `network_security_config.xml`(which will be applied to the app even if it doesn't have the file itself) that setting in the manifest will be ignored. We can review our network config but I would also suggest you to add a config file to the app instead of the manifest property, which would be a more secure and faster solution for the time being. Other than this I don't think there is anything specific about `axios` or installing the plugin.
> 
> It would be great if you could give this a try and let us know if it fixes the issue or not.

If we are nor adding the `network_security_config.xml` file, there is no issue with metro bundler.

Thanks.

**Comment by NMGuner** [comment]
Not exactly, metro bundler can work with the new config, the issue I linked has solutions for that in particular. Alongside with your development server info you also need to allow cleartext for localhost and emulator host device as in comments or the linked PR. Even without them, I would still suggest you to have a network security config file, at least in release, in order to only allow cleartext for certain domains as officially suggested by Google, instead of the deprecated manifest property which is vulnerable. 

As for the SDK, I can't promise if we can or will remove the file. That will need an investigation and will take some time, hence the suggestion for updating to the latest configuration type on your end.

**Comment by NMGuner** [comment]
Hi! We just released a new version (2.1.10) which includes the Android SDK with removed network security config. We hope that this could help resolve issues in your builds, if not please free to notify me here again and re-open the issue.
